File: src/math/AABB.ts
import { vec3, mat4, Vec3, Mat4 } from 'wgpu-matrix';

export class AABB {
  public min: Vec3;
  public max: Vec3;
  public corners: Vec3[];
  
  public minWS: Vec3;
  public maxWS: Vec3;
  public cornersWS: Vec3[];

  constructor() {
    this.min = vec3.fromValues(Infinity, Infinity, Infinity);
    this.max = vec3.fromValues(-Infinity, -Infinity, -Infinity);
    this.corners = this.computeCorners();
    
    this.minWS = vec3.fromValues(Infinity, Infinity, Infinity);
    this.maxWS = vec3.fromValues(-Infinity, -Infinity, -Infinity);
    this.cornersWS = this.computeCorners();
  }

  private computeCorners(): Vec3[] {
    return [
      vec3.create(), vec3.create(), vec3.create(), vec3.create(),
      vec3.create(), vec3.create(), vec3.create(), vec3.create(),
    ];
  }

  public setFromVertices(positions: number[]): void {
    for (let i = 0; i < positions.length; i += 3) {
      const x = positions[i];
      const y = positions[i + 1];
      const z = positions[i + 2];
      
      if (x < this.min[0]) this.min[0] = x;
      if (y < this.min[1]) this.min[1] = y;
      if (z < this.min[2]) this.min[2] = z;
      
      if (x > this.max[0]) this.max[0] = x;
      if (y > this.max[1]) this.max[1] = y;
      if (z > this.max[2]) this.max[2] = z;
    }

    this.updateCorners();
  }

  private updateCorners(): void {
    const min = this.min;
    const max = this.max;
    
    const c0 = this.corners[0]; c0[0] = min[0]; c0[1] = min[1]; c0[2] = min[2];
    const c1 = this.corners[1]; c1[0] = max[0]; c1[1] = min[1]; c1[2] = min[2];
    const c2 = this.corners[2]; c2[0] = max[0]; c2[1] = max[1]; c2[2] = min[2];
    const c3 = this.corners[3]; c3[0] = min[0]; c3[1] = max[1]; c3[2] = min[2];
    const c4 = this.corners[4]; c4[0] = min[0]; c4[1] = min[1]; c4[2] = max[2];
    const c5 = this.corners[5]; c5[0] = max[0]; c5[1] = min[1]; c5[2] = max[2];
    const c6 = this.corners[6]; c6[0] = max[0]; c6[1] = max[1]; c6[2] = max[2];
    const c7 = this.corners[7]; c7[0] = min[0]; c7[1] = max[1]; c7[2] = max[2];
  }

  public updateWorldSpace(worldMatrix: Mat4): void {
    this.minWS[0] = Infinity; this.minWS[1] = Infinity; this.minWS[2] = Infinity;
    this.maxWS[0] = -Infinity; this.maxWS[1] = -Infinity; this.maxWS[2] = -Infinity;

    for (let i = 0; i < 8; i++) {
      const corner = this.corners[i];
      const cornerWS = this.cornersWS[i];
      
      const x = corner[0], y = corner[1], z = corner[2];
      cornerWS[0] = worldMatrix[0] * x + worldMatrix[4] * y + worldMatrix[8] * z + worldMatrix[12];
      cornerWS[1] = worldMatrix[1] * x + worldMatrix[5] * y + worldMatrix[9] * z + worldMatrix[13];
      cornerWS[2] = worldMatrix[2] * x + worldMatrix[6] * y + worldMatrix[10] * z + worldMatrix[14];
      
      for (let j = 0; j < 3; j++) {
        if (cornerWS[j] < this.minWS[j]) this.minWS[j] = cornerWS[j];
        if (cornerWS[j] > this.maxWS[j]) this.maxWS[j] = cornerWS[j];
      }
    }
  }
}

File: references/rust_renderer_reference/primitives/aabb.rs
use engine_maths::{Const, OPoint, Point3};

use crate::primitives::global_transform::GlobalTransform;

#[derive(Debug, Clone)]
pub struct AABB {
    // Model space
    pub min: Point3<f32>,
    pub max: Point3<f32>,
    pub corners: [Point3<f32>; 8],
    // World space
    pub min_ws: Point3<f32>,
    pub max_ws: Point3<f32>,
    pub corners_ws: [Point3<f32>; 8],
    // Desc
    pub frustum_cull: bool,
}

impl AABB {
    pub fn new_from_vertex_positions(vertex_positions: &Vec<OPoint<f32, Const<3>>>, frustum_cull: bool) -> Self {
        let mut min = Point3::new(f32::INFINITY, f32::INFINITY, f32::INFINITY);
        let mut max = Point3::new(f32::NEG_INFINITY, f32::NEG_INFINITY, f32::NEG_INFINITY);

        for pos in vertex_positions {
            min.x = min.x.min(pos.x);
            min.y = min.y.min(pos.y);
            min.z = min.z.min(pos.z);

            max.x = max.x.max(pos.x);
            max.y = max.y.max(pos.y);
            max.z = max.z.max(pos.z);
        }

        let corners = [
            Point3::new(min.x, min.y, min.z),
            Point3::new(max.x, min.y, min.z),
            Point3::new(max.x, max.y, min.z),
            Point3::new(min.x, max.y, min.z),
            Point3::new(min.x, min.y, max.z),
            Point3::new(max.x, min.y, max.z),
            Point3::new(max.x, max.y, max.z),
            Point3::new(min.x, max.y, max.z),
        ];

        AABB {
            min,
            max,
            corners,
            min_ws: min,         // ws updated during render by the engine system
            max_ws: max,         // ws updated during render by the engine system
            corners_ws: corners, // ws updated during render by the engine system
            frustum_cull,
        }
    }

    /// Updates the AABB ws using a GlobalTransform
    pub fn update(&mut self, global_transform: &GlobalTransform) {
        let mut min_ws = Point3::new(f32::INFINITY, f32::INFINITY, f32::INFINITY);
        let mut max_ws = Point3::new(f32::NEG_INFINITY, f32::NEG_INFINITY, f32::NEG_INFINITY);

        for corner in &self.corners {
            let corner_ws = global_transform.matrix().transform_point(corner);

            min_ws.x = min_ws.x.min(corner_ws.x);
            min_ws.y = min_ws.y.min(corner_ws.y);
            min_ws.z = min_ws.z.min(corner_ws.z);

            max_ws.x = max_ws.x.max(corner_ws.x);
            max_ws.y = max_ws.y.max(corner_ws.y);
            max_ws.z = max_ws.z.max(corner_ws.z);
        }

        let corners_ws = [
            Point3::new(min_ws.x, min_ws.y, min_ws.z),
            Point3::new(max_ws.x, min_ws.y, min_ws.z),
            Point3::new(max_ws.x, max_ws.y, min_ws.z),
            Point3::new(min_ws.x, max_ws.y, min_ws.z),
            Point3::new(min_ws.x, min_ws.y, max_ws.z),
            Point3::new(max_ws.x, min_ws.y, max_ws.z),
            Point3::new(max_ws.x, max_ws.y, max_ws.z),
            Point3::new(min_ws.x, max_ws.y, max_ws.z),
        ];

        self.min_ws = min_ws;
        self.max_ws = max_ws;
        self.corners_ws = corners_ws;
    }
}
